streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -271,30 +270,40 @@ private long processNext(final List<ConsumerRecord<byte[], byte[]>> records,
                              final StateRestorer restorer,
                              final Long endOffset) {
         final List<KeyValue<byte[], byte[]>> restoreRecords = new ArrayList<>();
-        long offset = -1;
-
+        long nextPosition = -1;
+        int numberRecords = records.size();
+        int numberRestored = 0;
         for (final ConsumerRecord<byte[], byte[]> record : records) {
-            offset = record.offset();
+            final long offset = record.offset();
             if (restorer.hasCompleted(offset, endOffset)) {
+                nextPosition = record.offset();
                 break;
             }
+            numberRestored++;
             if (record.key() != null) {
                 restoreRecords.add(KeyValue.pair(record.key(), record.value()));
             }
         }
 
-        if (offset == -1) {
-            offset = restoreConsumer.position(restorer.partition());
+
+        // if we have changelog topic then we should have restored all records in the list
+        // otherwise if we did not fully restore to that point we need to set nextPosition
+        // to the position of the restoreConsumer and we'll cause a TaskMigratedException exception
+        if (nextPosition == -1 || (restorer.offsetLimit() == Long.MAX_VALUE && numberRecords != numberRestored)) {
```

Comment:
I've thought about this some more, and given the previous comments (https://github.com/apache/kafka/pull/4300#discussion_r156816065) that `processNext` was correct but susceptible to a race condition where we don't restore all records, IMHO we just need to add an additional check we did not restore all records in the list.   If you insist I can put in the `Math.min(restorer.OffestLimit(),restoreConsumer...)`

Reply:
I think this logic is correct now.`Math.min(...)` should do the exact same computation if I got it write. Both approaches have advantages: yours is more descriptive and thus explains itself better. The other one is more concise code allowing us to remove some local vars (`numberRecrods`, `numberRestored`), but harder to understand whats going if if you are not familiar with the code. Not sure which one is better.

@guozhangwang @dguy WDYT?